lib/collection/src/shards/replica_set.rs

```diff
@@ -687,6 +688,114 @@ impl ShardReplicaSet {
         captured_error.expect("at this point `captured_error` must be defined by construction")
     }
 
+    pub async fn execute_and_resolve_read_operation<'a, F, Fut, Res>(
+        &self,
+        read_operation: F,
+        local: &'a Option<Shard>,
+        remotes: &'a [RemoteShard],
+        factor: usize,
+        condition: ResolveCondition,
+    ) -> CollectionResult<Res>
+    where
+        F: Fn(&'a (dyn ShardOperation + Send + Sync)) -> Fut,
+        Fut: Future<Output = CollectionResult<Res>>,
+        Res: Resolve,
+    {
+        let local_count = usize::from(local.is_some());
+        let remotes_count = remotes.len();
+
+        let active_local = local
+            .as_ref()
+            .filter(|_| self.peer_is_active(&self.this_peer_id()));
+
+        let active_remotes_iter = remotes
+            .iter()
+            .filter(|remote| self.peer_is_active(&remote.peer_id));
+
+        let active_local_count = usize::from(active_local.is_some());
+        let active_remotes_count = active_remotes_iter.clone().count();
+
+        let total_count = local_count + remotes_count;
+        let active_count = active_local_count + active_remotes_count;
+
+        let factor = factor.clamp(1, total_count);
+
+        if active_count < factor {
+            return Err(CollectionError::service_error(format!(
+                "The replica set for shard {} on peer {} does not have enough active replicas",
+                self.shard_id,
+                self.this_peer_id(),
+            )));
+        }
+
+        let mut active_remotes: Vec<_> = active_remotes_iter.collect();
+        active_remotes.shuffle(&mut rand::thread_rng());
+
+        let local_operations = active_local
+            .into_iter()
+            .map(|local| read_operation(local.get()).left_future());
+
+        let remote_operations = active_remotes
+            .into_iter()
+            .map(|remote| read_operation(remote).right_future());
+
+        let mut operations = local_operations.chain(remote_operations);
+
+        let mut pending_operations: FuturesUnordered<_> = operations
+            .by_ref()
+            .take(factor + usize::try_from(self.read_remote_replicas).unwrap())
```

Comment:
```suggestion
            .take(factor + self.read_remote_replicas as usize)
```
?

Reply:
I always use `try_from` for `usize` conversions "just in case". E.g., on 32-bit platforms, there's a moderate chance that u64-as-usize will overflow and cause a bug that would be annoying to debug. In case of u32-as-usize that's a lot less likely, ofc.